Player update

victorbstan opened the pull request:
### Why?
Player code needs improvement

### What?
* Camera was in awkward position
* WIP Swimming code (note: liquid types implemented in another PR)

victorbstan committed:
Update .gitignore

victorbstan committed:
update camera transform in relation to container

victorbstan committed:
WIP player swim update

victorbstan committed:
player camera transform cleanup

victorbstan committed:
PLAYER cleanup

gongpha commented:
Can you elaborate on "Camera was in awkward position" ?
Your change seems to make the player looks shorter than the original.

Comparisons (can be observed at the position of the nails on the wall) :

The original (vkQuake)

https://github.com/gongpha/gdQmapbsp/assets/13400398/a218407c-9956-47de-81c6-cb854d882f0c

Your changes

https://github.com/gongpha/gdQmapbsp/assets/13400398/887e91fa-9322-41cf-9fa0-626937a93a86

Before the changes

https://github.com/gongpha/gdQmapbsp/assets/13400398/c44dd397-a6ef-4b14-bd27-04eb48abe7b2

And the swimming code you've noted as WIP looks OK. But it didn't work as expected yet.

https://github.com/gongpha/gdQmapbsp/assets/13400398/80a02c8a-f0ce-45ae-b93c-25a5f884dfe9

victorbstan committed:
removed gotdot git plugin

victorbstan committed:
Merge branch 'master' into player-update

victorbstan committed:
fix stairs walking while in liquid

victorbstan committed:
damage sounds setup, drowning process, some cleanup & refactoring liquids

victorbstan committed:
updated liquid damage processing

victorbstan committed:
updated step and ceiling checks

victorbstan commented:
> 

* Reset the camera height or "around. 

* Updated Step and Ceiling checks to have the step checking collider align with bottom of player and top of player + step height. This fixes an issue with not detecting very small  offsets on ground surface as in the following screenshots

<img width="1800" alt="Screenshot 2023-06-28 at 9 36 20 AM" src="https://github.com/gongpha/gdQmapbsp/assets/315297/f82d560a-376b-45a3-baed-af2ad6a4921b">
<img width="1800" alt="Screenshot 2023-06-28 at 9 19 49 AM" src="https://github.com/gongpha/gdQmapbsp/assets/315297/7deb0db8-9bea-4ed0-9266-9965339412c1">

* Fixed underwater movement to detect steps when in liquid
* Added mechanism to play a variety of player sounds for each liquid type 
* TODO: detect level of underwater submersion in order to process amount of hurt to apply, and if the player is drowning (currently value is hardcoded to "fully submerged"

victorbstan committed:
detect liquid submersion level, gasping for air sound

victorbstan commented:
* Added ability to detect player underwater submersion level: bellow mid-level = 1, above mid-level = 2, above eye-level = 3
* Drowning / damage in liquid multiplied by submersion level
* Coming up for air (gasping) sound added

victorbstan commented on quake1_example/scene/player.tscn:
```diff
@@ -14,23 +14,21 @@ collision_layer = 3
 
 [node name="around" type="Node3D" parent="."]
-transform = Transform3D(1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0.688, 0)
+transform = Transform3D(1, 0, 0, 0, 1, 0, 0, 0, 1, 2.08165e-12, 0.688, 2.08165e-12)
```
`2.08165e-12` shows up as `0` in editor, maybe this is a godot bug, or floating point rounding? 
<img width="273" alt="Screenshot 2023-06-28 at 4 45 06 PM" src="https://github.com/gongpha/gdQmapbsp/assets/315297/e655b2f8-09af-47b9-a639-4fdcac74a267">

gongpha replied:
Likely the latter. imo, that number seems very precise enough to me.

victorbstan replied:
I tried updating by typing `0` in the editor but it doesn't want to actually set it to `0` just keeps the odd number...

gongpha replied:
Confirmed that this is an unusual behavior. But occurred on the Y field instead
![image](https://github.com/gongpha/gdQmapbsp/assets/13400398/c5caa217-fe0a-4514-9d9e-af72b85194ab)

gongpha commented:
Haven't you added water sounds, right ?

and it'll be better if you'd fix the collision box when a player is trying to swim up to a water surface. I would appreciate it ^^

https://github.com/gongpha/gdQmapbsp/assets/13400398/db21055d-7e53-4269-a1d4-eba9f515d9c5

victorbstan commented:
> 

No water enter sounds (splashing yet, but that would be easy to add).

The player is currently able to bounce on top of the water, by jumping or swimming up, which is not ideal, but no worse than before.

I think as an improvement, perhaps the player should not be able to swim/jump up more than half the bbox height out of water. But that would also mean that the there needs to be updates to how "getting out of water" when near a step or solid object -- perhaps by increasing the valid step height while in liquid to include the total height of the player?

gongpha commented:
After my investigation. Let's say there are 2 crucial height points to check :

- Point A :
  - Approximately half of the player's bbox
  - Check if the player's upper half body is inside the water.
- Point B :
  - Roughly at the player's feet
  - Check if the player's feet are inside the water

![image](https://github.com/gongpha/gdQmapbsp/assets/13400398/3882d58b-dd06-45c0-bf79-b01009ad86ac)

Both points can be a very short box or something similar and **must touch the water to be considered as they're moving in water.**

See how :

https://github.com/gongpha/gdQmapbsp/assets/13400398/d46d2587-b0da-4bd4-91d7-9c08d6ed1a51

In the clip, You can see the player would fall out immediately when their feet are out of the water.

---

### Why needed Point B ? Point A alone will do stuff anyways.

Check out E4M2.

https://github.com/gongpha/gdQmapbsp/assets/13400398/b31b888b-11a5-4a54-bee4-23fbd4dcc5e7

The pool in E4M2 has a very thin water volume. The player couldn't even move in the water volume, despite Point A colliding with it. This is because their feet are already out of the water, making Point B doesn't collide.

Or even if Point B collides with water (their feet are inside the water) if Point A (their upper half body) doesn't collide with the water. It wouldn't be considered that they're moving in the water.

And sure, our implementation could do. which it shouldn't.

https://github.com/gongpha/gdQmapbsp/assets/13400398/d7c32266-a311-4ef7-b0d8-2ba9d513a171

This is my pure assumption. But should work in this situation :)

gongpha commented:
Also, I found a good source about the Quake metrics
https://book.leveldesignbook.com/process/blockout/metrics/quake

So, I will write it some of them here. And include Godot units too.

|Metric|Quake unit|Godot unit (1/32)|
|-|-|-|
|Collision hull size|32 x 32 x 56|1 x 1 x 1.75 (current is 1.5)
|camera height above floor|46|1.4375|
|Maximum run speed|320u/sec|10u/s|
|Fall damage height (for handling fall damage sound)|256+|8+|
|**Underwater height (above floor)**|29+|0.90625+|

victorbstan commented:
We can use `func _get_submerged_level()` to check 3 levels, bellow mid-point (bbox entered fluid volume) is `1`, mid-point (half point of character bbox) is `2`, and eye level (camera/around) is 3.

victorbstan commented:
There's also something we can decide to do, is ask if we want to implement "all the features", or work in a more incremental way, where we can have more peace-meal player improvements across many PR's. I don't know if we will successfully implement all the player features at once. 

Also I will be away this weekend, so won't touch this until later Monday or Tuesday.

victorbstan commented:
To make it more like in the original behaviour `func _get_submerged_level()` can be modified to add another point collision check at the foot level which would represent A and return 1... 

Personally I don't know if I would consider that an improvement, modifying behaviour to swim underwater when BBox touches water && mid-level (point B) touches water, would be good enough for me. But we can definitely try to be more like the original if we want...

victorbstan committed:
Merge branch 'master' into player-update

victorbstan committed:
fixed getting out of water, added check at feet level and made swimming behave more like traditional quake behaviour

victorbstan committed:
formatting

victorbstan commented:
Swimming should feel more like traditional game now (maybe not exactly, but closer)
Also getting out of water is easier now, as there's a separate "stairstep_liquid" which applies while in liquids, that is much higher than normal stairstep.

gongpha commented:
Looks better !
Also, I suggest you change the gravity as well. It seems a bit low. Compared with the original which falls faster.

it's around 800 or 25 in Godot unit.

victorbstan committed:
Merge branch 'master' into player-update

victorbstan committed:
updates after master merge to fix liquid collisions

victorbstan committed:
Merge branch 'master' into player-update

hemebond commented:
Is there an issue with ramps in this PR? I'm testing it on e1m1 and this 45 degree ramp can not be walked up:

![image](https://github.com/user-attachments/assets/5da4a5c6-1977-4ba2-b65e-1e8b209bab35)
